File: soap/Q2222_CharacterSlug.java
import java.text.Normalizer;
import soap.q2221.generated.CharacterService;
import soap.q2221.generated.SoapCharacterService;

public class Q2222_CharacterSlug {
    static final String STUDENT_CODE = "B21DCCN001";   // TODO
    static final String Q_CODE       = "TODO_qCode";   // TODO

    public static void main(String[] args) {
        CharacterService service = new CharacterService();
        SoapCharacterService port = service.getSoapCharacterServicePort();

        String data = port.requestString(STUDENT_CODE, Q_CODE);
        System.out.println("data=" + data);

        // lowercase → bỏ dấu Unicode → bỏ ký tự không phải chữ/số/space → gom space → dash
        String s = data.toLowerCase();
        s = Normalizer.normalize(s, Normalizer.Form.NFD)
                      .replaceAll("\\p{InCombiningDiacriticalMarks}+", "");
        s = s.replaceAll("[^a-z0-9\\s]", "");   // bỏ dấu câu
        s = s.trim().replaceAll("\\s+", " ");   // gom space
        String slug = s.replace(' ', '-');
        System.out.println("slug=" + slug);

        port.submitString(STUDENT_CODE, Q_CODE, slug);
        System.out.println("Đã submit slug=" + slug);
    }
}

File: soap/Q2222_CharacterSlug.py
import re
import unicodedata
from zeep import Client

EXAM_IP      = "36.50.135.242"
STUDENT_CODE = "B22DCCN863"   # TODO
Q_CODE       = "TODO_qCode"   # TODO
WSDL         = f"http://{EXAM_IP}:2221/CharacterService?wsdl"


def slugify(text: str) -> str:
    # lowercase → bỏ dấu Unicode (NFD + strip combining marks)
    s = text.lower()
    s = unicodedata.normalize("NFD", s)
    s = "".join(ch for ch in s if not unicodedata.combining(ch))
    # bỏ ký tự không phải chữ/số/space
    s = re.sub(r"[^a-z0-9\s]", "", s)
    # gom space → 1 space → đổi sang dash
    s = re.sub(r"\s+", " ", s).strip()
    return s.replace(" ", "-")


def main():
    client = Client(WSDL)

    data = client.service.requestString(STUDENT_CODE, Q_CODE)
    print("data=", data)

    slug = slugify(data)
    print("slug=", slug)

    client.service.submitString(STUDENT_CODE, Q_CODE, slug)
    print("Đã submit slug=", slug)


if __name__ == "__main__":
    main()
